improve messages for unused and too few args errors in std.fmt

xdBronch opened the pull request:
currently the error from code like this
```zig
print("{}\n\"\x00", .{ 0, 0 });
```
yields this error
```
zig/lib/std/fmt.zig:202:18: error: unused argument in '{}
                                               "
            1 => @compileError("unused argument in '" ++ fmt ++ "'"),
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
this looks terrible and hurts readability, after this pr it gives
```
zig/lib/std/fmt.zig:202:18: error: unused argument in '{}\n\"\x00'
            1 => @compileError(comptimePrint("unused argument in '{}'", .{std.zig.fmtEscapes(fmt)})),
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
which matches the format string exactly

i feel like i remember there being an issue about this but i couldnt find anything, it couldve just been a random conversation i read. if it exists and someone can find it that would be wonderful :pray:

rootbeer commented:
LGTM. Great to see improvements to the fmt error cases.  They can be annoying to map back the actual code.

I see `test/cases/compile_errors/std.fmt_error_for_unused_arguments.zig` has a related test case, may be worth following that example to add tests for the new behavior.

If you need ideas,

xdBronch commented:
thanks, yeah i saw that file earlier. its the only std test case in there which is a little weird imo but i guess fmt has lots of `@compileError`s. im not sure it would be all that helpful to add a test case for this change since its just making edge cases behave more like how they shouldve from the start, not really new behavior, at least the way i see it

xdBronch commented:
i realized that there really wasnt much point not testing this so i adjusted a test to account for this change and added a similar change + test for the opposite case.

andrewrk commented on lib/std/fmt.zig:
```diff
@@ -179,7 +180,7 @@ pub fn format(
 
         const arg_to_print = comptime arg_state.nextArg(arg_pos) orelse
-            @compileError("too few arguments");
+            @compileError(comptimePrint("too few arguments in '{}'", .{fmtEscapes(fmt)}));
```
This is too crazy, calling formatted printing inside the error message for bad formatting printing, which is all just to work around a limitation of compiler error messages when there is a comptime call stack.

It reeks of C++.

This is a great example of climbing up the wrong mountain.

Instead let us address the real problems:
- compile errors for formatted printing are not good enough *because of the compiler's inadequacy, not the std lib*
- `@compileError` should be the one to render unprintables correctly
- some kind of comptime-equivalent of `@returnAddress()` is needed

xdBronch replied:
you have a great point, the thought absolutely crossed my mind when making this pr, this *shouldnt* be necessary. the problem is that, as of now, i believe it is necessary. zig in general is not known for having friendly or descriptive error messages and `std.fmt` is notorious for bad ones. unless someone takes it upon themselves to improve the compiler enough to supersede this pr soon i think it would be a mistake not having it in some form or another. whats the path forward?

Vexu replied:
Also related https://github.com/ziglang/zig/issues/12093#issuecomment-1585340480

xdBronch replied:
something like that would be wonderful, one limitation of my pr here is that it doesnt have access to the actual in-source format string so if someone for example did `print("{}\x0a", .{})` my error would report back `"{}\n"` since `\x0a` is the hex representation of a line feed

andrewrk commented:
Closing, wrong direction for the project to take. Let us direct our efforts towards the root problems.

xdBronch commented:
im disappointed but i understand, the fix/improvement being in the compiler itself will be much better

andrewrk commented:
Thanks for understanding. If I don't do this, people will pile even more hacks on top of your code. I've seen it happen
